# -------------------------------- DNA Classification Project ---------------------------------------

## About :
### In this project, we will explore the world of bioinformatics by using Markov models, K-nearest neighbor (KNN) algorithms, support vector machines, and other common classifiers to classify short E. Coli DNA sequences. This project will use a dataset from the UCI Machine Learning Repository that has 106 DNA sequences, with 57 sequential nucleotides (“base-pairs”) each.





#### It includes :
<ul>
    <li>Importing data from the UCI repository</li>
    <li>Converting text inputs to numerical data</li>
    <li>Building and training classification algorithms</li>
    <li>Comparing and contrasting classification algorithms</li>
</ul>

In [ ]:
# Hide warnings
import warnings
warnings.simplefilter('ignore')

## Step 1: Importing the Dataset

The following code cells will import necessary libraries and import the dataset from the UCI repository as a Pandas DataFram

In [ ]:
#import and change module name
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/molecular-biology/splice-junction-gene-sequences/splice.data'

data = pd.read_csv(url, sep=',\s+', engine = 'python',header=None)
data.columns = names = ['Class', 'id', 'Sequence']

In [ ]:
data.head()

Class                  id                                           Sequence
0    EI    ATRINS-DONOR-521  CCAGCTGCATCACAGGAGGCCAGCGAGCAGGTCTGTTCCAAGGGCC...
1    EI    ATRINS-DONOR-905  AGACCCGCCGGGAGGCGGAGGACCTGCAGGGTGAGCCCCACCGCCC...
2    EI    BABAPOE-DONOR-30  GAGGTGAAGGACGTCCTTCCCCAGGAGCCGGTGAGAAGCGCAGTCG...
3    EI   BABAPOE-DONOR-867  GGGCTGCGTTGCTGGTCACATTCCTGGCAGGTATGGGGCGGGGCTT...
4    EI  BABAPOE-DONOR-2817  GCTCAGCCCCCAGGTCACCCAGGAACTGACGTGAGTGTCCCCATCC...

In [ ]:
data.columns

Index(['Class', 'id', 'Sequence'], dtype='object')

In [ ]:
data.head()

Class                  id                                           Sequence
0    EI    ATRINS-DONOR-521  CCAGCTGCATCACAGGAGGCCAGCGAGCAGGTCTGTTCCAAGGGCC...
1    EI    ATRINS-DONOR-905  AGACCCGCCGGGAGGCGGAGGACCTGCAGGGTGAGCCCCACCGCCC...
2    EI    BABAPOE-DONOR-30  GAGGTGAAGGACGTCCTTCCCCAGGAGCCGGTGAGAAGCGCAGTCG...
3    EI   BABAPOE-DONOR-867  GGGCTGCGTTGCTGGTCACATTCCTGGCAGGTATGGGGCGGGGCTT...
4    EI  BABAPOE-DONOR-2817  GCTCAGCCCCCAGGTCACCCAGGAACTGACGTGAGTGTCCCCATCC...

In [ ]:
data['Sequence'].head()

0    CCAGCTGCATCACAGGAGGCCAGCGAGCAGGTCTGTTCCAAGGGCC...
1    AGACCCGCCGGGAGGCGGAGGACCTGCAGGGTGAGCCCCACCGCCC...
2    GAGGTGAAGGACGTCCTTCCCCAGGAGCCGGTGAGAAGCGCAGTCG...
3    GGGCTGCGTTGCTGGTCACATTCCTGGCAGGTATGGGGCGGGGCTT...
4    GCTCAGCCCCCAGGTCACCCAGGAACTGACGTGAGTGTCCCCATCC...
Name: Sequence, dtype: object

In [ ]:
data.shape

(3190, 3)

In [ ]:
data.dtypes

Class       object
id          object
Sequence    object
dtype: object

## Step 2: Preprocessing the Dataset

The data is not in a usable form; as a result, we will need to process it before using it to train our algorithms.

In [ ]:
# Build our dataset using custom pandas dataframe
clases = data.loc[:,'Class']
clases.head()

0    EI
1    EI
2    EI
3    EI
4    EI
Name: Class, dtype: object

In [ ]:
# generate list of DNA sequence
sequence = list(data.loc[:,'Sequence'])
sequence

['CCAGCTGCATCACAGGAGGCCAGCGAGCAGGTCTGTTCCAAGGGCCTTCGAGCCAGTCTG',
 'AGACCCGCCGGGAGGCGGAGGACCTGCAGGGTGAGCCCCACCGCCCCTCCGTGCCCCCGC',
 'GAGGTGAAGGACGTCCTTCCCCAGGAGCCGGTGAGAAGCGCAGTCGGGGGCACGGGGATG',
 'GGGCTGCGTTGCTGGTCACATTCCTGGCAGGTATGGGGCGGGGCTTGCTCGGTTTTCCCC',
 'GCTCAGCCCCCAGGTCACCCAGGAACTGACGTGAGTGTCCCCATCCCGGCCCTTGACCCT',
 'CAGACTGGGTGGACAACAAAACCTTCAGCGGTAAGAGAGGGCCAAGCTCAGAGACCACAG',
 'CCTTTGAGGACAGCACCAAGAAGTGTGCAGGTACGTTCCCACCTGCCCTGGTGGCCGCCA',
 'CCCTCGTGCGGTCCACGACCAAGACCAGCGGTGAGCCACGGGCAGGCCGGGGTCGTGGGG',
 'TGGCGACTACGGCGCGGAGGCCCTGGAGAGGTGAGGACCCTCCTGTCCCTGCTCCAGTCC',
 'AAGCTGACAGTGGACCCGGTCAACTTCAAGGTGAGCCAGGAGTCGGGTGGGAGGGTGAGA',
 'TGGCGACTACGGCGCGGAGGCCCTGGAGAGGTGAGGACCCTGGTATCCCTGCTGCCAGTC',
 'AAGCTGAGAGTGGACCCTGTCAACTTCAAGGTGAGCCACCAGTCGGGTGGGGAGGGTGAG',
 'GGAAGATGCTGGAGGAGAAACCCTGGGAAGGTAGGCTCTGGTGACCAGGACAAGGGAGGG',
 'AAGCTGCATGTGGATCCTGAGAACTTCAGGGTGAGTACAGGAGATGTTTCAGCCCTGTTG',
 'GGAAGATGTTGGAGGAGAAACCCTGGGAAGGTAGGCTCTGGTGACCAGGACAAGGGAGGG',
 'AAGCTGCATGTGGATCCTGAGAA

In [ ]:
#Remove tab from each sequence
dic = {}
for i, seq in enumerate(sequence):
    nucleotides = list(seq)
    nucleotides = [char for char in nucleotides if char != '\t']
    #append class assignment
    nucleotides.append(clases[i])
    
    dic[i] = nucleotides
dic[0]    

['C',
 'C',
 'A',
 'G',
 'C',
 'T',
 'G',
 'C',
 'A',
 'T',
 'C',
 'A',
 'C',
 'A',
 'G',
 'G',
 'A',
 'G',
 'G',
 'C',
 'C',
 'A',
 'G',
 'C',
 'G',
 'A',
 'G',
 'C',
 'A',
 'G',
 'G',
 'T',
 'C',
 'T',
 'G',
 'T',
 'T',
 'C',
 'C',
 'A',
 'A',
 'G',
 'G',
 'G',
 'C',
 'C',
 'T',
 'T',
 'C',
 'G',
 'A',
 'G',
 'C',
 'C',
 'A',
 'G',
 'T',
 'C',
 'T',
 'G',
 'EI']

In [ ]:
# Convert Dict object into dataframe
df = pd.DataFrame(dic)
df.head()


0    1    2    3    4    5    6    7    8    9     ... 3180 3181 3182 3183  \
0    C    A    G    G    G    C    C    C    T    A  ...    C    C    T    T   
1    C    G    A    G    C    A    C    C    G    A  ...    C    T    G    A   
2    A    A    G    G    T    G    T    C    G    G  ...    T    G    T    A   
3    G    C    G    C    C    A    T    T    C    C  ...    G    G    T    A   
4    C    C    T    T    A    C    T    C    G    T  ...    C    A    T    A   

  3184 3185 3186 3187 3188 3189  
0    A    T    G    T    A    A  
1    T    C    A    C    T    G  
2    C    T    G    T    T    G  
3    A    C    C    C    C    C  
4    A    T    T    G    T    T  

[5 rows x 3190 columns]

In [ ]:
# transpose dataframe into correct format
df = df.transpose()
df.head()

0  1  2  3  4  5  6  7  8  9   ... 51 52 53 54 55 56 57 58 59  60
0  C  C  A  G  C  T  G  C  A  T  ...  G  C  C  A  G  T  C  T  G  EI
1  A  G  A  C  C  C  G  C  C  G  ...  T  G  C  C  C  C  C  G  C  EI
2  G  A  G  G  T  G  A  A  G  G  ...  A  C  G  G  G  G  A  T  G  EI
3  G  G  G  C  T  G  C  G  T  T  ...  G  T  T  T  T  C  C  C  C  EI
4  G  C  T  C  A  G  C  C  C  C  ...  C  T  T  G  A  C  C  C  T  EI

[5 rows x 61 columns]

In [ ]:
df.columns

RangeIndex(start=0, stop=61, step=1)

In [ ]:
# Rename
df.rename(columns = {60:'Class'}, inplace = True)

In [ ]:
df.columns

Index([      0,       1,       2,       3,       4,       5,       6,       7,
             8,       9,      10,      11,      12,      13,      14,      15,
            16,      17,      18,      19,      20,      21,      22,      23,
            24,      25,      26,      27,      28,      29,      30,      31,
            32,      33,      34,      35,      36,      37,      38,      39,
            40,      41,      42,      43,      44,      45,      46,      47,
            48,      49,      50,      51,      52,      53,      54,      55,
            56,      57,      58,      59, 'Class'],
      dtype='object')

In [ ]:
df.head()

0  1  2  3  4  5  6  7  8  9  ... 51 52 53 54 55 56 57 58 59 Class
0  C  C  A  G  C  T  G  C  A  T  ...  G  C  C  A  G  T  C  T  G    EI
1  A  G  A  C  C  C  G  C  C  G  ...  T  G  C  C  C  C  C  G  C    EI
2  G  A  G  G  T  G  A  A  G  G  ...  A  C  G  G  G  G  A  T  G    EI
3  G  G  G  C  T  G  C  G  T  T  ...  G  T  T  T  T  C  C  C  C    EI
4  G  C  T  C  A  G  C  C  C  C  ...  C  T  T  G  A  C  C  C  T    EI

[5 rows x 61 columns]

In [ ]:
#Encoding
numerical_df = pd.get_dummies(df)
numerical_df['cat']=df['Class']
numerical_df=numerical_df.drop(['Class_EI','Class_IE','Class_N'],axis=1)
cleanup_nums = {"cat":     {"EI": 1, "IE": 2, "N":3}}
numerical_df=numerical_df.replace(cleanup_nums)
numerical_df.head()

0_A  0_C  0_D  0_G  0_T  1_A  1_C  1_D  1_G  1_T  ...  58_C  58_G  58_N  \
0    0    1    0    0    0    0    1    0    0    0  ...     0     0     0   
1    1    0    0    0    0    0    0    0    1    0  ...     0     1     0   
2    0    0    0    1    0    1    0    0    0    0  ...     0     0     0   
3    0    0    0    1    0    0    0    0    1    0  ...     1     0     0   
4    0    0    0    1    0    0    1    0    0    0  ...     1     0     0   

   58_T  59_A  59_C  59_G  59_N  59_T  cat  
0     1     0     0     1     0     0    1  
1     0     0     1     0     0     0    1  
2     1     0     0     1     0     0    1  
3     0     0     1     0     0     0    1  
4     0     0     0     0     0     1    1  

[5 rows x 288 columns]

In [ ]:
# rename Class_+ to Class
numerical_df.rename(columns = {'cat':'Class'}, inplace = True)

## Step 3: Training and Testing the Classification Algorithms

Now that we have preprocessed the data and built our training and testing datasets, we can start to deploy different classification algorithms. It's relatively easy to test multiple models; as a result, we will compare and contrast the performance of ten different algorithms.

In [ ]:
#Importing different classifier from sklearn
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn import svm
from sklearn.naive_bayes import GaussianNB
from sklearn.gaussian_process.kernels import RBF
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.metrics import classification_report, accuracy_score

In [ ]:
from sklearn.model_selection import train_test_split
X = numerical_df.drop(['Class'], axis = 1).values
y = numerical_df['Class'].values

#define a seed for reproducibility
seed = 1

# Splitting data into training and testing data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = seed)


In [ ]:
# Define scoring method
scoring = 'accuracy'
# Model building to train
names = ['K Nearest Neighbors', 'Gaussian Process', 'Decision Tree', 'Random Forest', 'Neural Net', 'AddaBoost', 'Naive Bayes', 'SVM Linear', 'SVM RBF', 'SVM Sigmoid']
Classifiers = [
    KNeighborsClassifier(n_neighbors = 3),
    DecisionTreeClassifier(max_depth = 5),
    RandomForestClassifier(max_depth = 5, n_estimators = 10, max_features = 1 ),
    MLPClassifier(alpha = 1),
    AdaBoostClassifier(),
    GaussianNB(),
    svm.SVC(kernel = 'linear'),
    svm.SVC(kernel = 'rbf'),
    svm.SVC(kernel = 'sigmoid')
    
    ]
models = zip(names, Classifiers)
# import KFold
from sklearn.model_selection import KFold, cross_val_score

names = []
result = []
for name, model in models:
    kfold = KFold(n_splits = 10, random_state = 1,shuffle=True)
    cv_results = cross_val_score(model, X_train, y_train, cv = kfold, scoring = 'accuracy')
    result.append(cv_results)
    names.append(name)
    msg = "{0}: {1} ({2})".format(name, cv_results.mean(), cv_results.std())
    print(msg)

K Nearest Neighbors: 0.7592119944211995 (0.023434473735693767)
Gaussian Process: 0.9327004881450488 (0.01213620559874672)
Decision Tree: 0.5229654811715482 (0.02894803796198816)
Random Forest: 0.952339609483961 (0.011398167157503828)
Neural Net: 0.9280892608089262 (0.018040726980547005)
AddaBoost: 0.8934013249651324 (0.021785515365652618)
Naive Bayes: 0.9222437238493726 (0.019442084379450948)
SVM Linear: 0.9631973500697351 (0.010742416965548743)
SVM RBF: 0.948573919107392 (0.008192901810110054)


## Step 4 : Model Evaluation

Now that we will evaluate our classification algorithms using accuracy score and classification report.

In [ ]:
#Test the algorithm on the test data set
models = zip(names, Classifiers)
for name, model in models:
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print(name)
    print(accuracy_score(y_test, y_pred))
    print(classification_report(y_test, y_pred))
    

K Nearest Neighbors
0.7769423558897243
              precision    recall  f1-score   support

           1       0.62      0.93      0.74       191
           2       0.76      0.89      0.82       198
           3       0.95      0.65      0.78       409

    accuracy                           0.78       798
   macro avg       0.78      0.82      0.78       798
weighted avg       0.82      0.78      0.78       798

Gaussian Process
0.9461152882205514
              precision    recall  f1-score   support

           1       0.94      0.94      0.94       191
           2       0.89      0.94      0.91       198
           3       0.98      0.95      0.96       409

    accuracy                           0.95       798
   macro avg       0.94      0.94      0.94       798
weighted avg       0.95      0.95      0.95       798

Decision Tree
0.5175438596491229
              precision    recall  f1-score   support

           1       0.00      0.00      0.00       191
           2       1.

## Conclusion : 

#### <i>From above report, Support Vector Machine with 'linear' kernel performed best with F1_score  = 0.96 on testing data.</i>

### Thanks !